### DISCUSSION TITLE
WithObjectUnchanged

### DISCUSSION THREAD

**Discussion by o-nnerb** [pointfreeco/swift-composable-architecture#2193] [Ideas]
Adopting a view like `WithObjectUnchanged` could be used to prevent rebuilding the ViewStore when the initial State remains the same.

The ideia is to add the Equatable conformance to Reducer.State so we could use `WithObjectUnchanged` to verify if updating the viewStore is truly necessary.

```
public struct WithObjectUnchanged<Input, Content, Object>: View where Input: Hashable, Content: View {

    public init(
        _ input: Input,
        _ object: @escaping (Input) -> Object,
        @ViewBuilder content: @escaping (Object) -> Content
    )
```

This way, we could have something like this:

```
var body: some View {
    WithObjectUnchanged(scene, \.store) {
        WithViewStore($0) {
            MovieView(viewStore: $0)
        }
    }
}
```

With this approach, the block initializing the ViewStore is executed only once, regardless of how many times the body is called. The view will only be rebuilt if the input, in this case the Scene property, is updated with a new value.

**Comment by mbrandonw** [comment]
Hi @brennobemoura, this probably isn't something we will bring into the library. It adds another layer of indentation, and it's yet another concept to learn.

In the end we are moving in a [direction](https://twitter.com/pointfreeco/status/1667248728862871612) that allows us to get rid of the view store entirely (and `WithViewStore`) and so these problems will eventually (hopefully) be a thing of the past.

**Reply by o-nnerb** [reply]
Yeah, you're right. Also, after hearing some people share their point of view, it became clear to me that this wouldn't be the right path for the TCA. 🙃
